File: plugins/plugin-im/model/group.go
package model

import "time"

// ─── Group Models ──────────────────────────────────────────────────────

// Group 群组
type Group struct {
	ID         string     `gorm:"primaryKey;size:32" json:"id"`
	Name       string     `gorm:"size:100;not null" json:"name"`
	Avatar     string     `gorm:"size:255" json:"avatar"`
	OwnerID    string     `gorm:"size:32;not null;index" json:"owner_id"`
	OwnerType  string     `gorm:"size:20;not null" json:"owner_type"`
	GroupType  string     `gorm:"size:20;not null;default:mixed" json:"group_type"`
	Notice     string     `gorm:"type:text" json:"notice"`
	IsPublic   bool       `gorm:"default:false" json:"is_public"`
	MaxMembers int        `gorm:"default:200" json:"max_members"`
	Status     string     `gorm:"size:10;not null;default:normal" json:"status"`
	CreatedAt  *time.Time `json:"created_at"`
	UpdatedAt  *time.Time `json:"updated_at"`
}

func (Group) TableName() string { return "im_group" }

// GroupMember 群成员
type GroupMember struct {
	ID         string     `gorm:"primaryKey;size:32" json:"id"`
	GroupID    string     `gorm:"size:32;not null;uniqueIndex:idx_gm_group_user;index" json:"group_id"`
	UserID     string     `gorm:"size:32;not null;uniqueIndex:idx_gm_group_user" json:"user_id"`
	UserType   string     `gorm:"size:20;not null;uniqueIndex:idx_gm_group_user" json:"user_type"`
	Role       string     `gorm:"size:10;not null;default:member" json:"role"`
	Nickname   string     `gorm:"size:100" json:"nickname"`
	MutedUntil *time.Time `json:"muted_until"`
	JoinedAt   *time.Time `json:"joined_at"`
	Status     string     `gorm:"size:10;not null;default:active" json:"status"`
}

func (GroupMember) TableName() string { return "im_group_member" }

// GroupJoinRequest 入群申请
type GroupJoinRequest struct {
	ID       string     `gorm:"primaryKey;size:32" json:"id"`
	GroupID  string     `gorm:"size:32;not null;index" json:"group_id"`
	UserID   string     `gorm:"size:32;not null" json:"user_id"`
	UserType string     `gorm:"size:20;not null" json:"user_type"`
	Reason   string     `gorm:"size:255" json:"reason"`
	Status   string     `gorm:"size:10;not null;default:pending" json:"status"` // pending | approved | rejected
	HandledBy string    `gorm:"size:32" json:"handled_by"`
	CreatedAt *time.Time `json:"created_at"`
	UpdatedAt *time.Time `json:"updated_at"`
}

func (GroupJoinRequest) TableName() string { return "im_group_join_request" }

// GroupMessage 群消息
type GroupMessage struct {
	ID         string     `gorm:"primaryKey;size:32" json:"id"`
	GroupID    string     `gorm:"size:32;not null;index:idx_gmsg_group_created,priority:1" json:"group_id"`
	SenderID   string     `gorm:"size:32;not null;index" json:"sender_id"`
	SenderType string     `gorm:"size:20;not null" json:"sender_type"`
	Content    string     `gorm:"type:text" json:"content"`
	Extra      string     `gorm:"type:text" json:"extra"`
	MsgType    string     `gorm:"size:20;not null;default:text" json:"msg_type"`
	ReplyTo    string     `gorm:"size:32;index" json:"reply_to"`
	CreatedAt  *time.Time `gorm:"index:idx_gmsg_group_created,priority:2" json:"created_at"`
}

func (GroupMessage) TableName() string { return "im_group_message" }

// GroupMessageRead 群消息已读
type GroupMessageRead struct {
	ID        string     `gorm:"primaryKey;size:32" json:"id"`
	MessageID string     `gorm:"size:32;not null;uniqueIndex:idx_gmr_msg_user" json:"message_id"`
	GroupID   string     `gorm:"size:32;not null;index" json:"group_id"`
	UserID    string     `gorm:"size:32;not null;uniqueIndex:idx_gmr_msg_user" json:"user_id"`
	UserType  string     `gorm:"size:20;not null;uniqueIndex:idx_gmr_msg_user" json:"user_type"`
	ReadAt    *time.Time `json:"read_at"`
}

func (GroupMessageRead) TableName() string { return "im_group_message_read" }

File: plugins/plugin_im/model/group.py
"""Group chat ORM models."""

from __future__ import annotations

from datetime import datetime
from typing import Optional
from sqlalchemy import Index, text
from sqlalchemy.dialects.mysql import BOOLEAN, DATETIME, INTEGER, TEXT, VARCHAR
from sqlalchemy.orm import Mapped, mapped_column
from sdk.kernel.registry import HeiBase


class Group(HeiBase):
    """群组"""
    __tablename__ = "im_group"
    __table_args__ = (
        Index("idx_owner_id", "owner_id"),
    )

    id: Mapped[str] = mapped_column(VARCHAR(32, charset="utf8mb4", collation="utf8mb4_general_ci"), primary_key=True)
    name: Mapped[str] = mapped_column(VARCHAR(100, charset="utf8mb4", collation="utf8mb4_general_ci"), nullable=False)
    avatar: Mapped[str] = mapped_column(VARCHAR(255, charset="utf8mb4", collation="utf8mb4_general_ci"), default="", server_default=text("''"))
    owner_id: Mapped[str] = mapped_column(VARCHAR(32, charset="utf8mb4", collation="utf8mb4_general_ci"), nullable=False)
    owner_type: Mapped[str] = mapped_column(VARCHAR(20, charset="utf8mb4", collation="utf8mb4_general_ci"), nullable=False)
    group_type: Mapped[str] = mapped_column(VARCHAR(20, charset="utf8mb4", collation="utf8mb4_general_ci"), nullable=False, default="mixed", server_default=text("'mixed'"))
    notice: Mapped[str] = mapped_column(TEXT(collation="utf8mb4_general_ci"), default="", server_default=text("''"))
    is_public: Mapped[bool] = mapped_column(BOOLEAN, default=False, server_default=text("false"))
    max_members: Mapped[int] = mapped_column(INTEGER, default=200, server_default=text("200"))
    status: Mapped[str] = mapped_column(VARCHAR(10, charset="utf8mb4", collation="utf8mb4_general_ci"), nullable=False, default="normal", server_default=text("'normal'"))
    created_at: Mapped[Optional[datetime]] = mapped_column(DATETIME)
    updated_at: Mapped[Optional[datetime]] = mapped_column(DATETIME)


class GroupMember(HeiBase):
    """群成员"""
    __tablename__ = "im_group_member"
    __table_args__ = (
        Index("idx_gm_group_user", "group_id", "user_id", "user_type", unique=True),
        Index("idx_group_id", "group_id"),
    )

    id: Mapped[str] = mapped_column(VARCHAR(32, charset="utf8mb4", collation="utf8mb4_general_ci"), primary_key=True)
    group_id: Mapped[str] = mapped_column(VARCHAR(32, charset="utf8mb4", collation="utf8mb4_general_ci"), nullable=False)
    user_id: Mapped[str] = mapped_column(VARCHAR(32, charset="utf8mb4", collation="utf8mb4_general_ci"), nullable=False)
    user_type: Mapped[str] = mapped_column(VARCHAR(20, charset="utf8mb4", collation="utf8mb4_general_ci"), nullable=False)
    role: Mapped[str] = mapped_column(VARCHAR(10, charset="utf8mb4", collation="utf8mb4_general_ci"), nullable=False, default="member", server_default=text("'member'"))
    nickname: Mapped[str] = mapped_column(VARCHAR(100, charset="utf8mb4", collation="utf8mb4_general_ci"), default="", server_default=text("''"))
    muted_until: Mapped[Optional[datetime]] = mapped_column(DATETIME)
    joined_at: Mapped[Optional[datetime]] = mapped_column(DATETIME)
    status: Mapped[str] = mapped_column(VARCHAR(10, charset="utf8mb4", collation="utf8mb4_general_ci"), nullable=False, default="active", server_default=text("'active'"))


class GroupJoinRequest(HeiBase):
    """加群请求"""
    __tablename__ = "im_group_join_request"
    __table_args__ = (
        Index("idx_group_id", "group_id"),
    )

    id: Mapped[str] = mapped_column(VARCHAR(32, charset="utf8mb4", collation="utf8mb4_general_ci"), primary_key=True)
    group_id: Mapped[str] = mapped_column(VARCHAR(32, charset="utf8mb4", collation="utf8mb4_general_ci"), nullable=False)
    user_id: Mapped[str] = mapped_column(VARCHAR(32, charset="utf8mb4", collation="utf8mb4_general_ci"), nullable=False)
    user_type: Mapped[str] = mapped_column(VARCHAR(20, charset="utf8mb4", collation="utf8mb4_general_ci"), nullable=False)
    reason: Mapped[str] = mapped_column(VARCHAR(255, charset="utf8mb4", collation="utf8mb4_general_ci"), default="", server_default=text("''"))
    status: Mapped[str] = mapped_column(VARCHAR(10, charset="utf8mb4", collation="utf8mb4_general_ci"), nullable=False, default="pending", server_default=text("'pending'"))
    handled_by: Mapped[str] = mapped_column(VARCHAR(32, charset="utf8mb4", collation="utf8mb4_general_ci"), default="", server_default=text("''"))
    created_at: Mapped[Optional[datetime]] = mapped_column(DATETIME)
    updated_at: Mapped[Optional[datetime]] = mapped_column(DATETIME)


class GroupMessage(HeiBase):
    """群消息"""
    __tablename__ = "im_group_message"
    __table_args__ = (
        Index("idx_gmsg_group_created", "group_id", "created_at"),
        Index("idx_sender_id", "sender_id"),
        Index("idx_reply_to", "reply_to"),
    )

    id: Mapped[str] = mapped_column(VARCHAR(32, charset="utf8mb4", collation="utf8mb4_general_ci"), primary_key=True)
    group_id: Mapped[str] = mapped_column(VARCHAR(32, charset="utf8mb4", collation="utf8mb4_general_ci"), nullable=False)
    sender_id: Mapped[str] = mapped_column(VARCHAR(32, charset="utf8mb4", collation="utf8mb4_general_ci"), nullable=False)
    sender_type: Mapped[str] = mapped_column(VARCHAR(20, charset="utf8mb4", collation="utf8mb4_general_ci"), nullable=False)
    content: Mapped[str] = mapped_column(TEXT(collation="utf8mb4_general_ci"))
    extra: Mapped[str] = mapped_column(TEXT(collation="utf8mb4_general_ci"))
    msg_type: Mapped[str] = mapped_column(VARCHAR(20, charset="utf8mb4", collation="utf8mb4_general_ci"), nullable=False, default="text", server_default=text("'text'"))
    reply_to: Mapped[str] = mapped_column(VARCHAR(32, charset="utf8mb4", collation="utf8mb4_general_ci"), default="", server_default=text("''"))
    created_at: Mapped[Optional[datetime]] = mapped_column(DATETIME)


class GroupMessageRead(HeiBase):
    """群消息已读记录"""
    __tablename__ = "im_group_message_read"
    __table_args__ = (
        Index("idx_gmr_msg_user", "message_id", "user_id", "user_type", unique=True),
        Index("idx_group_id", "group_id"),
    )

    id: Mapped[str] = mapped_column(VARCHAR(32, charset="utf8mb4", collation="utf8mb4_general_ci"), primary_key=True)
    message_id: Mapped[str] = mapped_column(VARCHAR(32, charset="utf8mb4", collation="utf8mb4_general_ci"), nullable=False)
    group_id: Mapped[str] = mapped_column(VARCHAR(32, charset="utf8mb4", collation="utf8mb4_general_ci"), nullable=False)
    user_id: Mapped[str] = mapped_column(VARCHAR(32, charset="utf8mb4", collation="utf8mb4_general_ci"), nullable=False)
    user_type: Mapped[str] = mapped_column(VARCHAR(20, charset="utf8mb4", collation="utf8mb4_general_ci"), nullable=False)
    read_at: Mapped[Optional[datetime]] = mapped_column(DATETIME)
